src/s2/util/math/exactfloat/bignum.cc

```diff
@@ -0,0 +1,855 @@
+      nzero += Bignum::kBigitBits;
+    } else {
+      nzero += absl::countr_zero(bigit);
+      break;
+    }
+  }
+  return nzero;
+}
+
+bool Bignum::is_bit_set(int nbit) const {
+  ABSL_DCHECK_GE(nbit, 0);
+  const size_t digit = nbit / kBigitBits;
+  const size_t shift = nbit % kBigitBits;
+
+  if (digit >= bigits_.size()) {
+    return false;
+  }
+
+  return ((bigits_[digit] >> shift) & 0x1) != 0;
+}
+
+Bignum Bignum::operator-() const {
+  Bignum result = *this;
+  result.negate();
+  return result;
+}
+
+Bignum& Bignum::operator<<=(int nbit) {
+  ABSL_DCHECK_GE(nbit, 0);
+  if (is_zero() || nbit == 0) {
+    return *this;
+  }
+
+  const int nbigit = nbit / kBigitBits;
+  const int nrem = nbit % kBigitBits;
+
+  // First, handle the whole-bigit shift by inserting zeros.
+  bigits_.insert(bigits_.begin(), nbigit, 0);
+
+  // Then, handle the within-bigit shift, if any.
+  if (nrem != 0) {
+    Bigit carry = 0;
+    for (size_t i = 0; i < bigits_.size(); ++i) {
+      const Bigit old_val = bigits_[i];
+      bigits_[i] = (old_val << nrem) | carry;
+      carry = old_val >> (kBigitBits - nrem);
+    }
+
+    if (carry) {
+      bigits_.push_back(carry);
+    }
+  }
+
+  return *this;
+}
+
+Bignum& Bignum::operator>>=(int nbit) {
+  ABSL_DCHECK_GE(nbit, 0);
+  if (is_zero() || nbit == 0) {
+    return *this;
+  }
+
+  // Shifting by more than the bit width results in zero.
+  if (nbit >= bit_width(*this)) {
+    return set_zero();
+  }
+
+  const int nbigit = nbit / kBigitBits;
+  const int nrem = nbit % kBigitBits;
+
+  // First, handle the whole-bigit shift by removing bigits.
+  bigits_.erase(bigits_.begin(), bigits_.begin() + nbigit);
+
+  // Then, handle the within-bigit shift, if any.
+  if (nrem != 0) {
+    Bigit carry = 0;
+    for (auto i = static_cast<int64_t>(bigits_.size()) - 1; i >= 0; --i) {
+      const Bigit old_val = bigits_[i];
+      bigits_[i] = (old_val >> nrem) | carry;
+      carry = old_val << (kBigitBits - nrem);
+    }
+  }
+
+  // Result might be smaller or zero, so normalize.
+  Normalize();
+  return *this;
+}
+
+// Raise this value to the given power, which must be non-negative.
+Bignum Bignum::Pow(int32_t pow) const {
+  ABSL_DCHECK_GE(pow, 0);
+
+  // Anything to the zero-th power is 1 (including zero).
+  if (pow == 0) {
+    return Bignum(1);
+  }
+
+  if (is_zero()) {
+    return Bignum(0);
+  }
+
+  // Core algorithm: Exponentiation by squaring.
+  Bignum result(1);
+  Bignum base = *this;  // A mutable copy of the base.
+  uint32_t upow = static_cast<uint32_t>(pow);
+
+  while (upow > 0) {
+    if (upow & 1) {  // If current exponent bit is 1, multiply into result.
+      result *= base;
+    }
+    base *= base;
+    upow >>= 1;
+  }
+
+  return result;
+}
+
+// Computes a + b + carry and updates the carry.
+inline Bigit AddBigit(Bigit a, Bigit b, Bigit* absl_nonnull carry) {
+#ifdef __x86_64__
+  Bigit out;
+  *carry =
+      _addcarry_u64(*carry, a, b, reinterpret_cast<unsigned long long*>(&out));
```

Comment:
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_addcarry_u64&ig_expand=175
says `__int64`, so I would use that (assuming the header is actually written with that type).

```c++
static_assert(sizeof(Bigit) == sizeof(__int64));
```